### ISSUE TITLE
support Core elements substituted into text() strings

### ISSUE THREAD

**Issue by villebro** [sqlalchemy/sqlalchemy#4502]
Not sure if this is a feature request or a question, but is there a way to generate a custom expression, similarly to `sqlalchemy.func`, which is able to honor column quoting rules made by the engine? For example, if I want to create the custom expression

```sql
DATEADD(MINUTE, FLOOR(DATE_PART(MINUTE, {col}) / 5) * 5, DATE_TRUNC('HOUR', {col}))
```

I currently have to make a `literal_column`, replacing the `{col}` with the column name to which I want to run the expression on. However, when doing so, the compiler isn't aware that `{col}`is actually a column. Is there a way of making the above using native sqlalchemy constructs? I've tried reading up on the documentation, but I was not able to understand how this could be done.

**Comment by zzzeek** [comment]
you would use [column](https://docs.sqlalchemy.org/en/latest/core/sqlelement.html?highlight=column#sqlalchemy.sql.expression.column) for this:

```
>>> from sqlalchemy import func, column
>>> print(func.DATE_THING(column('CaseSensitive')))
DATE_THING("CaseSensitive")
>>> print(func.DATE_THING(column('case_insensitive')))
DATE_THING(case_insensitive)

```

**Comment by villebro** [comment]
Thanks for the answer. For a simple case, such as the part with `DATE_PART(MINUTE, {col})`, doing something like `func.DATE_PART(literal_column('MINUTE'), column('CaseSensitive'))` seems feasible, but I'm not sure how I would go about writing the expression above using `func`. A really neat feature would be something where one could write the expression somehow similar to f-strings, like `func.expression('DATEADD(MINUTE, FLOOR(DATE_PART(MINUTE, {col}) / 5) * 5, DATE_TRUNC('HOUR', {col}))', col=column('CaseSensitive'))`.

**Comment by zzzeek** [comment]
for DATE_PART, we have support for this built in using EXTRACT: https://docs.sqlalchemy.org/en/latest/core/sqlelement.html?highlight=extract#sqlalchemy.sql.expression.extract   which unfortunately seems to be lacking documentation (?!) , it looks like some docs got deleted at some point, looks like: https://github.com/sqlalchemy/sqlalchemy/blob/master/test/sql/test_functions.py#L980

as far as string expressions, there is only one construct that is taking a string which is text: https://docs.sqlalchemy.org/en/latest/core/sqlelement.html?highlight=text#sqlalchemy.sql.expression.text     adding any additional functions that accept SQL strings is total non-starter as nowadays we get frivolous CVEs posted throughout all security trackers on the internet for any such functions and I just got finished removing a huge amount of textual coercion for 1.3.   

 text() accepts additional modifiers such as .bindparams() and .columns() so i will gladly review a PR that includes a new feature `.expressions(**kw)` which when present means that additional tokens in the string can be replaced with a dialect-compiled form of each token passed to this method.    When the method is not used, no such conversion takes place so as to provide for backwards compatibility for SQL strings that include unescaped brackets.    A system of escaping is also needed so that the brackets can still be used directly even when the .expressions feature is enabled.   The system would require a complete suite of tests added to test/sql/test_text.py.

**Comment by villebro** [comment]
Thanks for the comprehensive answer. I will research the docs/code to see if I could potentially start working on a PR for this.

**Comment by zzzeek** [comment]
im also curious if we should use brackets as the character, because would that annoy people who are trying to use f strings already?

**Comment by villebro** [comment]
You're absolutely right, the bracket proposal was just me shooting from the hip. I'm perfectly happy to settle for any character that is readable and doesn't clash with standard python syntax.

**Comment by zzzeek** [comment]
therein lies the challenge

**Comment by villebro** [comment]
Ping @victornoel if you still feel like solving https://github.com/apache/incubator-superset/issues/5886 elegantly. I will not be able to work on this for a few weeks, but feel tempted to take a stab at this, and can help out if you want to take the lead.

**Comment by victornoel** [comment]
@villebro hehe, thanks for the vote of confidence, but unfortunately I'm 100% sure I don't have the skills for this kind contribution :)

**Comment by villebro** [comment]
@zzzeek feel free to assign this to me if this isn't something that is high priority (will work on it when I have the time).

**Comment by villebro** [comment]
@zzzeek I now have a working prototype that has the functionality I'm looking for. Any thoughts before I start cleaning up the code, writing test cases and writing documentation?

```python
import sqlalchemy as sa
from sqlalchemy.sql import column, text
engine = sa.create_engine('oracle://uid:pwd@localhost/test')

a = text('select coalesce($col1, $col2) as col from tbl')\
    .expressions({'$col1': column('lower_case_col'), '$col2': column('MixedCaseCol')})
print(a.compile(engine))
```
which produces
```
select coalesce(lower_case_col, "MixedCaseCol") as col from tbl
```

**Comment by villebro** [comment]
Working on this I'm not entirely convinced `TextClause` is the correct placeholder for this functionality. My use case (and I expect others to have similar needs) I would ideally need something in between a `column`, `literal_column` and `func`, i.e. something that can be used inside a select statement and has all the same functionality, e.g. the `label` function. I propose adding a new function `expression_column` to complement `column` and `literal_column`, returning a `ColumnClause` which has an extra argument `keywords` expecting a dict, which then replaces those keywords in the original text. Example:

```python
import sqlalchemy as sa
from sqlalchemy.sql import column, expression_column, table, select

cols = {
    '$col1': column('lower_case_col'),
    '$col2': column('MixedCaseCol'),
    '$col3': column('decimal'),
}
expr = expression_column("COALESCE($col1, $col2, $col3)" , cols).label('MyAlias')
tbl = table('tbl')
sel = select([expr]).select_from(tbl)
```
Which for Oracle would generate
```sql
SELECT COALESCE(lower_case_col, "MixedCaseCol", "decimal") AS "MyAlias" FROM tbl
```
Thoughts?

**Comment by zzzeek** [comment]
hi @villebro -

I like what you're doing here, however I'm not sure what the demand for this feature is right now.  There's nothing to prevent you from making a very nice third party library that does this using `@compiles`.    I understand the rationale for putting the full token syntax in the dictionary (e.g. `{'${key};': 'value'}`) but I can see that not being so well received.

**Comment by villebro** [comment]
I think this is such a minor feature that it feels overkill to build a third party library for this. So will perhaps put this on the shelf for now. But if the need for this comes up in the future I don't mind putting in the work for it, as I think it would complement the existing functionality quite nicely.

**Comment by zzzeek** [comment]
ok.  I think you had something interesting here. I think to your issue above there is generally a need for a "text" that goes anywhere and a "literal_column()" that is specifically a SQL expression.   adding a third kind of element is not solving that problem, it's more like both literal_column() and text() would somehow do this thing you are looking for, or we'd have a better story for why text() doesn't work where literal_column() does (I think it generally *does* work, however, perhaps text() should be the basis for literal_column(), that is, text("foobar").as_column_expression(), something like that).

**Comment by villebro** [comment]
The hint about using `@compiles` was great, and upon closer inspection I'm thinking it should be possible to squeeze the logic I need in some 10-15 LOC. If that works out a more general solution might emerge. We'll see. Should I reopen this and post some code once it's done or open up a new issue? I'm pretty sure I need your guidance on where in the codebase this would fit, as I'm still fairly overwhelmed by the inner workings of the library.

**Comment by zzzeek** [comment]
I'd want it to fit within the existing constructs because if we add new constructs they will never be found.     ideally long term text() is the only way people type a SQL string in the library at all.

**Comment by villebro** [comment]
Is it your opinion that the syntax would be something along the lines of
`expr = text('coalesce($col1, $col2)', col1=column('snake_case'), col2=column('MixedCase'))`?

**Comment by zzzeek** [comment]
it would use method chaining, not kw args, e.g. `text(...).with_tokens(...)` something like that.

**Comment by villebro** [comment]
Yes, I agree. i will clean up my own code based on the compiles decorator to show my use case, you can then determine if there is a case for integrating it with into the main codebase.

**Comment by villebro** [comment]
Ok, here's my use case:

```python
from sqlalchemy.ext.compiler import compiles
from sqlalchemy.sql.expression import ColumnClause
from sqlalchemy.sql import column, literal_column, table, select

class TokenExpression(ColumnClause):
    def __init__(self, text, **kwargs):
        self.tokens = {}
        super().__init__(text, **kwargs)
    
    def with_tokens(self, **kwargs):
        for key, value in kwargs.items():
            self.tokens[key] = value
        return self

@compiles(TokenExpression)
def compile_token_expression(element, compiler, **kw):
    text = element.name
    for key, value in element.tokens.items():
        text = text.replace('$' + key, compiler.process(element.tokens[key], **kw))
    return text

token_col = TokenExpression('COALESCE($col1, $col2)').\
    with_tokens(col1=column('MixedCase'), col2=literal_column('MixedCase')).\
    label('abc')
tbl = table('tbl')
sel = select([token_col]).select_from(tbl)
print(sel)
```
This produces
```sql
SELECT COALESCE("MixedCase", MixedCase) AS abc 
FROM tbl
```
Obviously this would need escaping and other checks, but for my use case this works just fine. The point here is that I currently need to extend `ColumnClause`, as `TextClause` doesn't offer `label()`. But if `.as_column_expression()` were to be added to `TextClause`, then this would probably sit well there, too.

**Comment by zzzeek** [comment]
I suggest propose add `.label()` to `TextClause`, returns `literal_column(<text>).label()`.  The `ColumnClause` is what is returned by `literal_column`.   We can also add `.as_column()` to get just the `ColumnClause` part.  A new collection `_substitutions` is added to `TextClause` and `ColumnClause`, likely as part of a common mixin `Substitutable`.   Elements are added to the `'_substitutions` collection using the method `substitutions()`.    note that this also implicitly adds `literal_column(..).substitutions()` as well.

Two choices for the token:

1. use the same as bindparam, e.g. `:token`.   This would propose that we rewrite the system by which `TextClause` scans the string in `__init__` and builds `bindparam()` objects to instead keep each token open-ended as a placeholder, and the remaining ones become `bindparam()` objects inside of SQLCompiler, which would not be an unusual approach.    Technically, `substitutions()` does basically the same thing as `bindparam()`, that is, associates clause elements with string tokens, it's just `bindparams()` uses the `bindparam()` object.  So to that extent I like this version because it generalizes existing TextClause functionality and does not add new complexity.

2. hedge a bit and use either `::token` or `:token:`.   Using the colon is the only backwards compatible approach and I doubt anyone is naming their parameters with a colon at the front.   This would allow the SQL statement to visually indicate those places that are bound parameters vs. other kinds of expressions...although I'm not sure it really matters.  

Assuming #1 or even #2, the `_substittutions` collection can likely internally replace the `TextClause._bindparams` collection.    `SQLCompiler` moves out parts of its `visit_textclause` implementation that deals with the tokens into a common method that is used for the two `Substitutable` descendants (`TextClause` and `ColumnClause`).

The idea here is:

1. we add no new public facing classes 

2. text() remains as the single place that all textual SQL functionality is consolidated so it can all be used together

3. we add a single method that does the thing and it even acts as a superset for `.bindparams()`.

4. the feature gets added to `TextClause` and `literal_column` implicitly and also starts to remove the need for people to have to remember to use `literal_column` which really should be that they are using `TextClause`.

5. nothing is backwards incompatible


Basically if someone is writing SQL as a string, they should have to write the word `text()`.  It's just one place, with one well known set of caveats (e.g. do NOT put untrusted input into it) and one well known set of functionality (use `:token` liberally and substitute in any way you'd like with bindparam(), cast(), column(), func() or whatever.)

**Comment by villebro** [comment]
Took a while to digest this, but yeah I like the proposal and it covers all my needs. Re: the choices for the token, going with #1 is fine, as long as it doesn't interfere with or confuse users on how binds are used. So if we have :col1 in the text, and later do a where on the same token, a user might be confused as to what gets substituted where. But I assume you have a much better understanding of how this will work in practice. But I'm assuming that substitutions won't be exposed to regular binds (right?). I think adding more tokens, unless explicitly required, will just be confusing.

The general concept of the proposal are clear, but will require some more familiarization with the codebase ffor me to be able to complete the feature. Like I said I don't mind doing the work for this, but it might be slow, and will require some help along the way. So if you have a clear vision of how this should be carried out feel free to do the dirty work (my main motivation here is getting the functionality into the library, not a merged PR). Otherwise, I will put together a WIP PR that we can work on together until it's ok.

**Comment by villebro** [comment]
..either way I will be happy to help maintain the functionality going forward.

**Comment by zzzeek** [comment]
Yes it was necessarily a lot to digest as we have to put a lot of thought into every feature these days, but I can clarify any specific part of it as needed.    Feel free to take as much time as you'd like as there is no urgent need for this feature (but I do think it will be very nice).

**Comment by zzzeek** [comment]
also, if part of the proposal simply can't work, like the part where we change how _bindparams works, we might need to get into that.  I didn't work out every line of how that would change and there may be some behavioral contracts in test/sql/test_text.py that rely on its behavior.

**Comment by villebro** [comment]
Sounds like a plan. I will gradually start preparing a WIP PR.
